### PR TITLE
Add a warning when calling a QNode with non-tensor positional arguments

### LINKED ISSUE #1459: WireError being thrown whenever running circuits where wire names are given explicitly
I have recently run a clean install of Pennylane in a new environment, and now when I run circuits where the wires are provided explicitly, for example

```
dev = qml.device('default.qubit', wires=5)
@qml.qnode(dev)
def test(wires):
   qml.Hadamard(wires=wires[0])
   return qml.state()

wires=range(5)
test(wires)
```

I get the following error

```
File "<ipython-input-8-45aee5197fda>", line 1, in <module>
    test(wires)

  File "/home/abhi/anaconda3/envs/pennylane-env/lib/python3.9/site-packages/pennylane/qnode.py", line 598, in __call__
    res = self.qtape.execute(device=self.device)

  File "/home/abhi/anaconda3/envs/pennylane-env/lib/python3.9/site-packages/pennylane/tape/tape.py", line 1323, in execute
    return self._execute(params, device=device)

  File "/home/abhi/anaconda3/envs/pennylane-env/lib/python3.9/site-packages/pennylane/tape/tape.py", line 1354, in execute_device
    res = device.execute(self)

  File "/home/abhi/anaconda3/envs/pennylane-env/lib/python3.9/site-packages/pennylane/_qubit_device.py", line 192, in execute
    self.apply(circuit.operations, rotations=circuit.diagonalizing_gates, **kwargs)

  File "/home/abhi/anaconda3/envs/pennylane-env/lib/python3.9/site-packages/pennylane/devices/default_qubit.py", line 200, in apply
    self._state = self._apply_operation(self._state, operation)

  File "/home/abhi/anaconda3/envs/pennylane-env/lib/python3.9/site-packages/pennylane/devices/default_qubit.py", line 231, in _apply_operation
    return self._apply_unitary_einsum(state, matrix, wires)

  File "/home/abhi/anaconda3/envs/pennylane-env/lib/python3.9/site-packages/pennylane/devices/default_qubit.py", line 629, in _apply_unitary_einsum
    device_wires = self.map_wires(wires)

  File "/home/abhi/anaconda3/envs/pennylane-env/lib/python3.9/site-packages/pennylane/devices/default_qubit.py", line 168, in map_wires
    raise WireError(

WireError: Did not find some of the wires (tensor(0, requires_grad=True),) on device with wires (0, 1, 2, 3, 4).
```

This never happened previously. Can someone help me debug? 

My python version is 3.9.5, and here are the installed packages:

# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main
_openmp_mutex             4.5                       1_gnu
absl-py                   0.13.0                   pypi_0    pypi
alabaster                 0.7.12             pyhd3eb1b0_0
appdirs                   1.4.4                    pypi_0    pypi
argh                      0.26.2           py39h06a4308_0
argon2-cffi               20.1.0           py39h27cfd23_1
astor                     0.8.1                    pypi_0    pypi
astroid                   2.6.2            py39h06a4308_0
async_generator           1.10               pyhd3eb1b0_0
atomicwrites              1.4.0                      py_0
attrs                     21.2.0             pyhd3eb1b0_0
autograd                  1.3                      pypi_0    pypi
autopep8                  1.5.6              pyhd3eb1b0_0
autoray                   0.2.5                    pypi_0    pypi
babel                     2.9.1              pyhd3eb1b0_0
backcall                  0.2.0              pyhd3eb1b0_0
black                     20.8b1                   pypi_0    pypi
bleach                    3.3.0              pyhd3eb1b0_0
brotlipy                  0.7.0           py39h27cfd23_1003
ca-certificates           2021.7.5             h06a4308_1
cachetools                4.2.2                    pypi_0    pypi
certifi                   2021.5.30        py39h06a4308_0
cffi                      1.14.6           py39h400218f_0
chardet                   4.0.0           py39h06a4308_1003
charset-normalizer        2.0.2                    pypi_0    pypi
cirq                      0.11.0                   pypi_0    pypi
cirq-core                 0.11.0                   pypi_0    pypi
cirq-google               0.11.0                   pypi_0    pypi
click                     8.0.1              pyhd3eb1b0_0
cloudpickle               1.6.0                      py_0
cryptography              3.4.7            py39hd23ed53_0
cycler                    0.10.0                   pypi_0    pypi
dbus                      1.13.18              hb2f20db_0
decorator                 4.3.0                    pypi_0    pypi
defusedxml                0.7.1              pyhd3eb1b0_0
diff-match-patch          20200713                   py_0
docutils                  0.17.1           py39h06a4308_1
entrypoints               0.3              py39h06a4308_0
expat                     2.4.1                h2531618_2
flake8                    3.9.0              pyhd3eb1b0_0
flynt                     0.65                     pypi_0    pypi
fontconfig                2.13.1               h6c09931_0
freetype                  2.10.4               h5ab3b9f_0
future                    0.18.2           py39h06a4308_1
glib                      2.69.0               h5202010_0
google-api-core           1.31.0                   pypi_0    pypi
google-auth               1.33.0                   pypi_0    pypi
googleapis-common-protos  1.53.0                   pypi_0    pypi
graycode                  1.0.5                    pypi_0    pypi
grpcio                    1.38.1                   pypi_0    pypi
gst-plugins-base          1.14.0               h8213a91_2
gstreamer                 1.14.0               h28cd5cc_2
icu                       58.2                 he6710b0_3
idna                      3.2                      pypi_0    pypi
imagesize                 1.2.0              pyhd3eb1b0_0
importlib-metadata        3.10.0           py39h06a4308_0
importlib_metadata        3.10.0               hd3eb1b0_0
iniconfig                 1.1.1                    pypi_0    pypi
intervaltree              3.1.0                      py_0
ipykernel                 5.3.4            py39hb070fc8_0
ipython                   7.22.0           py39hb070fc8_0
ipython_genutils          0.2.0              pyhd3eb1b0_1
isort                     5.9.2              pyhd3eb1b0_0
jedi                      0.17.2           py39h06a4308_1
jeepney                   0.7.0              pyhd3eb1b0_0
jinja2                    3.0.1              pyhd3eb1b0_0
joblib                    1.0.1                    pypi_0    pypi
jpeg                      9b                   h024ee3a_2
jsonschema                3.2.0                      py_2
jupyter_client            6.1.12             pyhd3eb1b0_0
jupyter_core              4.7.1            py39h06a4308_0
jupyterlab_pygments       0.1.2                      py_0
keyring                   23.0.1           py39h06a4308_0
kiwisolver                1.3.1                    pypi_0    pypi
lazy-object-proxy         1.6.0            py39h27cfd23_0
ld_impl_linux-64          2.35.1               h7274673_9
libffi                    3.3                  he6710b0_2
libgcc-ng                 9.3.0               h5101ec6_17
libgomp                   9.3.0               h5101ec6_17
libpng                    1.6.37               hbc83047_0
libsodium                 1.0.18               h7b6447c_0
libspatialindex           1.9.3                h2531618_0
libstdcxx-ng              9.3.0               hd4cf53a_17
libuuid                   1.0.3                h1bed415_2
libxcb                    1.14                 h7b6447c_0
libxml2                   2.9.12               h03d6c58_0
markupsafe                2.0.1            py39h27cfd23_0
matplotlib                3.4.2                    pypi_0    pypi
mccabe                    0.6.1            py39h06a4308_1
mistune                   0.8.4           py39h27cfd23_1000
mpmath                    1.2.1                    pypi_0    pypi
mypy-extensions           0.4.3                    pypi_0    pypi
nbclient                  0.5.3              pyhd3eb1b0_0
nbconvert                 6.1.0            py39h06a4308_0
nbformat                  5.1.3              pyhd3eb1b0_0
ncurses                   6.2                  he6710b0_1
nest-asyncio              1.5.1              pyhd3eb1b0_0
networkx                  2.5.1                    pypi_0    pypi
notebook                  6.4.0            py39h06a4308_0
numpy                     1.21.0                   pypi_0    pypi
numpydoc                  1.1.0              pyhd3eb1b0_1
openssl                   1.1.1k               h27cfd23_0
packaging                 21.0               pyhd3eb1b0_0
pandas                    1.3.0                    pypi_0    pypi
pandoc                    2.12                 h06a4308_0
pandocfilters             1.4.3            py39h06a4308_1
parso                     0.7.0                      py_0
pathspec                  0.8.1                    pypi_0    pypi
pcre                      8.45                 h295c915_0
pennylane                 0.16.0                   pypi_0    pypi
pennylane-cirq            0.16.0                   pypi_0    pypi
pexpect                   4.8.0              pyhd3eb1b0_3
pickleshare               0.7.5           pyhd3eb1b0_1003
pillow                    8.3.1                    pypi_0    pypi
pip                       21.1.3           py39h06a4308_0
pluggy                    0.13.1           py39h06a4308_0
prometheus_client         0.11.0             pyhd3eb1b0_0
prompt-toolkit            3.0.17             pyh06a4308_0
protobuf                  3.13.0                   pypi_0    pypi
psutil                    5.8.0            py39h27cfd23_1
ptyprocess                0.7.0              pyhd3eb1b0_2
py                        1.10.0                   pypi_0    pypi
py-bobyqa                 1.3                      pypi_0    pypi
pyasn1                    0.4.8                    pypi_0    pypi
pyasn1-modules            0.2.8                    pypi_0    pypi
pybind11                  2.6.2                    pypi_0    pypi
pycodestyle               2.6.0              pyhd3eb1b0_0
pycparser                 2.20                       py_2
pydocstyle                6.1.1              pyhd3eb1b0_0
pyflakes                  2.2.0              pyhd3eb1b0_0
pygments                  2.9.0              pyhd3eb1b0_0
pylint                    2.9.3            py39h06a4308_1
pyls-black                0.4.6                hd3eb1b0_0
pyls-spyder               0.3.2              pyhd3eb1b0_0
pyopenssl                 20.0.1             pyhd3eb1b0_1
pyparsing                 2.4.7              pyhd3eb1b0_0
pyqt                      5.9.2            py39h2531618_6
pyrsistent                0.18.0           py39h7f8727e_0
pysocks                   1.7.1            py39h06a4308_0
pytest                    6.2.4                    pypi_0    pypi
python                    3.9.5                h12debd9_4
python-dateutil           2.8.1              pyhd3eb1b0_0
python-jsonrpc-server     0.4.0                      py_0
python-language-server    0.36.2             pyhd3eb1b0_0
pytz                      2021.1             pyhd3eb1b0_0
pyxdg                     0.27               pyhd3eb1b0_0
pyyaml                    5.4.1            py39h27cfd23_1
pyzmq                     20.0.0           py39h2531618_1
qdarkstyle                2.8.1                      py_0
qsimcirq                  0.9.5                    pypi_0    pypi
qt                        5.9.7                h5867ecd_1
qtawesome                 1.0.2              pyhd3eb1b0_0
qtconsole                 5.1.0              pyhd3eb1b0_0
qtpy                      1.9.0                      py_0
readline                  8.1                  h27cfd23_0
regex                     2021.7.6                 pypi_0    pypi
requests                  2.26.0                   pypi_0    pypi
rope                      0.19.0             pyhd3eb1b0_0
rsa                       4.7.2                    pypi_0    pypi
rtree                     0.9.7            py39h06a4308_1
scikit-learn              0.24.2                   pypi_0    pypi
scikit-quant              0.8.2                    pypi_0    pypi
scipy                     1.7.0                    pypi_0    pypi
secretstorage             3.3.1            py39h06a4308_0
semantic-version          2.6.0                    pypi_0    pypi
send2trash                1.5.0              pyhd3eb1b0_1
setuptools                52.0.0           py39h06a4308_0
sip                       4.19.13          py39h2531618_0
six                       1.16.0             pyhd3eb1b0_0
snowballstemmer           2.1.0              pyhd3eb1b0_0
sortedcontainers          2.4.0              pyhd3eb1b0_0
sphinx                    4.0.2              pyhd3eb1b0_0
sphinxcontrib-applehelp   1.0.2              pyhd3eb1b0_0
sphinxcontrib-devhelp     1.0.2              pyhd3eb1b0_0
sphinxcontrib-htmlhelp    2.0.0              pyhd3eb1b0_0
sphinxcontrib-jsmath      1.0.1              pyhd3eb1b0_0
sphinxcontrib-qthelp      1.0.3              pyhd3eb1b0_0
sphinxcontrib-serializinghtml 1.1.5              pyhd3eb1b0_0
spyder                    4.2.5            py39h06a4308_0
spyder-kernels            1.10.2           py39h06a4308_0
sqcommon                  0.3.2                    pypi_0    pypi
sqimfil                   0.3.7                    pypi_0    pypi
sqlite                    3.36.0               hc218d9a_0
sqsnobfit                 0.4.5                    pypi_0    pypi
sympy                     1.8                      pypi_0    pypi
terminado                 0.9.4            py39h06a4308_0
testpath                  0.5.0              pyhd3eb1b0_0
textdistance              4.2.1              pyhd3eb1b0_0
threadpoolctl             2.2.0                    pypi_0    pypi
three-merge               0.1.1              pyhd3eb1b0_0
tk                        8.6.10               hbc83047_0
toml                      0.10.2             pyhd3eb1b0_0
tornado                   6.1              py39h27cfd23_0
tqdm                      4.61.2                   pypi_0    pypi
traitlets                 5.0.5              pyhd3eb1b0_0
typed-ast                 1.4.3                    pypi_0    pypi
typing_extensions         3.10.0.0           pyh06a4308_0
tzdata                    2021a                h52ac0ba_0
ujson                     4.0.2            py39h2531618_0
urllib3                   1.26.6             pyhd3eb1b0_1
watchdog                  1.0.2            py39h06a4308_1
wcwidth                   0.2.5                      py_0
webencodings              0.5.1            py39h06a4308_1
wheel                     0.36.2             pyhd3eb1b0_0
wrapt                     1.12.1           py39he8ac12f_1
wurlitzer                 2.1.0            py39h06a4308_0
xz                        5.2.5                h7b6447c_0
yaml                      0.2.5                h7b6447c_0
yapf                      0.31.0             pyhd3eb1b0_0
zeromq                    4.3.4                h2531618_0
zipp                      3.5.0              pyhd3eb1b0_0
zlib                      1.2.11               h7b6447c_3

### PR DESCRIPTION
**Pull request by antalszava**
Closes #1459.

### CODE REVIEW

File: pennylane/qnode.py [hunk 1]
```diff
@@ -592,6 +592,16 @@ def construct(self, args, kwargs):
 
     def __call__(self, *args, **kwargs):
 
+        for pos_arg in args:
+            if not isinstance(pos_arg, qml.numpy.tensor):
+                warnings.warn(
+                    "At least one of the positional arguments passed to the QNode"
+                    "is not a PennyLane NumPy tensor, though it is regarded as a trainable parameter. "
+                    "Please revisit the positional arguments when calling the QNode and pass non-tensor "
+                    "arguments as keyword arguments when calling the QNode.",
```
Comment (josh146) [inline pennylane/qnode.py:601]: Suggest rewording this to clarify the change, and avoid mentions of tensors. Something like the following:

```python
f"""VisibleDeprecationWarning: Passing non-differentiable positional arguments to
QNodes without setting requires_grad=False has been deprecated and will be
removed from a future release.

Please specify directly that this argument is non-trainable via np.array({pos_arg}, requires_grad=False),
or for non-numeric arguments consider passing it as a keyword argument."""
```

File: pennylane/qnode.py [hunk 2]
```diff
@@ -592,6 +592,16 @@ def construct(self, args, kwargs):
 
     def __call__(self, *args, **kwargs):
 
+        for pos_arg in args:
+            if not isinstance(pos_arg, qml.numpy.tensor):
```
Comment (josh146) [inline pennylane/qnode.py:596]: This warning should only be raised when using the autograd interface, _not_ pytorch/jax/tf!

**Review by josh146 — CHANGES_REQUESTED** [review]
@antalszava this will cause a warning to be raised for behaviour that is already consider acceptable by PL!

E.g.,

```python
@qml.qnode(dev)
def circuit(x):
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0))

>>> circuit(0.2)
WARNING: 
0.5432543
>>> circuit(np.array(0.2, requires_grad=True))
0.5432543
```

Furthermore, this _only applies to Autograd_.

Hence, we need to be very careful here. I suggest:

1. Making this a VisibleDeprecationWarning (since we are deprecating current PL behaviour).
2. Avoid mentions of 'NumPy tensor', as I think this could be confusing to users.
2. Re-wording the warning to make it clear exactly the change the user has to make; either use `requires_grad=False`, or pass as a keyword argument.

File: .github/CHANGELOG.md [hunk 3]
```diff
@@ -258,6 +258,10 @@
 
 <h3>Improvements</h3>
 
+* Adds a warning when one of the positional arguments with which a QNode is
+  being called is not a tensor.
+  [(#1470)](https://github.com/PennyLaneAI/pennylane/pull/1470)
```
Comment (josh146) [inline .github/CHANGELOG.md:263]: This will need to go into breaking changes!

Comment (josh146) [reply .github/CHANGELOG.md:263]: (and re-worded to make the deprecation more explicit)

File: pennylane/qnode.py (hunk 2 above)
Comment (josh146) [reply pennylane/qnode.py:596]: In addition, I believe it should check:

1. Whether the positional argument requires grad (via `qml.math.requires_grad`)
2. Only then, if it is not a `np.tensor`, should the warning be raised.

**Comment by antalszava** [comment]
Closing this as we've made the deprecation another way.
